records/2021-09-27.batch-celery-save.md

```diff
@@ -0,0 +1,116 @@
+with a larger context.
+
+An individual commit in a database in not a problem usually, but when
+several of these occurs, potentially over 100,000 in the case of Notify in time
+of high traffic, it becomes a performance issue.
+
+For a recap, imagine you are at the grocery store. You get to teh cashier to pay
+your items. You have to pay for each these items individually, i.e.
+a) scan one and only one item, b) ask for the total, c) pay with cash or card,
+d) repeat for all remaining items. This will take quite a long time compared
+to paying for all items in one batch within one transaction. Saving one
+notification every time in the database is akin to paying for one grocery item
+individually in one transaction, except you have potentially thousands of items
+remaining afterward. Ideally, we want to buy/save all for all of these in one
+go.
+
+### Current notification flow through Celery tasks
+
+> **INFO**: [Celery](https://docs.celeryproject.org/en/stable/getting-started/introduction.html) is a Python library to process a message through a workflow of tasks which offers reliability mechanism (e.g. retry), monitoring, concurrency (to execute multiple work at the same time), automatic scheduling of tasks and allows for ways to scale up the processing.
+
+Our current workflow of Celery tasks for processing an email or SMS at the moment
+is simplified with the next diagram.
+
+![Current workflow of a notification save](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/2021-09-27.batch-celery-save/current-workflow.jpg?raw=true "Current workflow of a notification save")
+
+## Options
+
+Listing of potential solutions and necessary architectural changes will be listed in
+this section. Feel free to list your own idea in here via a proposed change to this
+document.
+
+### [celery-batches](https://github.com/clokep/celery-batches)
+
+Around the development of [Celery version 3](https://docs.celeryproject.org/en/3.1/reference/celery.contrib.batches.html), there was an option to batch the content
+of a queue and have a task flush it all when one of the two constraint would be met:
+the number of maximum requests or a timeout. For example, with a limit of 1000
+requests maximum and a timeout of 5 seconds, if 1,000 requets would be filled in 2
+seconds, then the task would trigger and flush all of the 1,000 requests buffered
+in the current task. Alternatively, if only 2,300 requests filled up and the 5
+seconds timeout completed, then all of the 2,3000 requests would get flushed to
+the task.
+
+This is a transparent and easy mechanism that could get leveraged in Celery in order
+to achieve the batch save. Unfortunately, that feature was always in an experimental
+state in Celery version 3 and disappeared in subsequent versions. An individual
+contributer on GitHub [made a fork of the experimental feature](https://github.com/clokep/celery-batches) and adapted for recent
```

Comment:
Is this `has adapted for` or just `adapted for` ?

Reply:
Fixed it with `has adapted the latter for...`.